feat(parser): add a TIFF parser using utif

zarov opened the pull request:
It automatically rejects TIFFs that are bigger than 4096x4096, because
[it is the minimum that is supported on 99.99% of the devices][0].

[0]: https://webglstats.com/webgl/parameter/MAX_TEXTURE_SIZE

zarov committed:
feat(parser): add a TIFF parser using utif

It automatically rejects TIFFs that are bigger than 4096x4096, because
it is the minimum that is supported on 99.99% of the devices.

https://github.com/photopea/UTIF.js
https://www.npmjs.com/package/utif

gchoqueux commented on src/Parser/TIFFParser.js:
```diff
@@ -0,0 +1,67 @@
+            for (let j = 0; j < rowSize; j += alpha) {
+                resizedData[rowOffset + j] = it.next().value;
+                resizedData[rowOffset + j + 1] = it.next().value;
+                resizedData[rowOffset + j + 2] = it.next().value;
```
the alpha isn't take in account?

zarov replied:
It is set to `255` before this loop, I didn't add it here because it needs a `if`.

gchoqueux commented on src/Parser/TIFFParser.js:
```diff
@@ -0,0 +1,67 @@
+        }
+
+        // Round to next power of two
+        const width = 2 ** (~~Math.log2(IFD.width - 1) + 1);
```
there is utils in THREE to compute `next power of two`.

zarov replied:
Nice, changed thanks !

gchoqueux commented on src/Parser/TIFFParser.js:
```diff
@@ -0,0 +1,67 @@
+        let rowOffset = 0;
+        for (let i = IFD.width; i > 0; i--) {
+            rowOffset = i * width * alpha;
+            resizedData.fill(255, rowOffset, rowOffset + rowSize);
```
why you fill the `resizedData`.

zarov replied:
For the opacity, see other comment

gchoqueux commented on src/Parser/TIFFParser.js:
```diff
@@ -0,0 +1,67 @@
+        // Avoid using alpha layer if the TIFF is already at the right size
+        const alpha = (IFD.width == width && IFD.height == height) ? 3 : 4;
+
+        const resizedData = new IFD.data.constructor(width * height * alpha);
```
`THREE` must be able to handle image resizing.

zarov replied:
Yes, but I don't want to stretch the image from its initial size, and I want to avoid having `THREE` warnings in the console.

gchoqueux commented on examples/tiff.html:
```diff
@@ -0,0 +1,74 @@
+                itowns.Fetcher.arrayBuffer('https://raw.githubusercontent.com/iTowns/iTowns2-sample-data/master/tiff/FLAG_T24.TIF')
+                    .then(itowns.TIFFParser.parse)
+                    .then(function _(texture) {
+                        texture.coords = new itowns.Coordinates('EPSG:4326', 1.4116667, 43.6074422);
```
The `texture.coords` have bad name, because it's actually a `Extent`. 
this Extent allows you to place the texture on the tiles. In particular with initialization from parent texture. 

There are little changes to enhance the image display. 
* Use `Extent` insteaf of `Coordinates`, and add it a `zoom` property to set image scale.
* Add a layer property : use `imageParent` after zoom max to always display the ColorLayer after the Source.zoom.max.

zarov replied:
This needs another change in [`MaterialLayer` as well](https://github.com/iTowns/itowns/blob/master/src/Renderer/MaterialLayer.js#L135).

> Add a layer property : use imageParent after zoom max to always display the ColorLayer after the Source.zoom.max.

I don't understand what you mean here

gchoqueux replied:
A ColorLayer isn't displayed when Source.zoom.max is exceeded, it isn't the expected behaviour because the ColorLayer displaying shouldn't be controlled by the source.

zarov commented:
Updated, with a tiled example (the same as https://www.itowns-project.org/demo-geoid/) but needs #1126 and #1142 to be merged first.

zarov commented:
Now that all others things have been merged, this PR is now ready !

gchoqueux commented on examples/tiff.html:
```diff
@@ -0,0 +1,53 @@
+
+            // Add one imagery layer to the scene, read from TIFFs.
+            var tiffSource = new itowns.TMSSource({
+                url: 'http://localhost/iTowns2-sample-data/geoid/localcolors/tiff/${z}/localcolors_${x}_${y}.tif',
```
could add the tiff on iTowns2-sample-data?

zarov replied:
Done

gchoqueux commented on src/Parser/TIFFParser.js:
```diff
@@ -0,0 +1,74 @@
+
+        // A 4096 limitation sounds the most probable here
+        // https://webglstats.com/webgl/parameter/MAX_TEXTURE_SIZE
+        if (IFD.width > 4096 || IFD.height > 4069) {
```
use gpu capabilities.

zarov replied:
Done

gchoqueux commented on src/Parser/TIFFParser.js:
```diff
@@ -0,0 +1,74 @@
+        if (width == IFD.width && height == IFD.height) {
+            resizedData = IFD.data;
+        } else {
+            resizedData = new IFD.data.constructor(width * height * 4);
```
THREE use canvas to resize image see : https://github.com/mrdoob/three.js/blob/master/src/renderers/webgl/WebGLTextures.js#L63

zarov replied:
Yes but then we got a warning: I want to avoid that.

gchoqueux replied:
yes, but may be is better, to resizing, to use `canva.context.drawImage( image, 0, 0, width, height )`.

zarov replied:
I tried it: it is too complicated IMO, because `drawImage` can't take a `TypedArray` as an argument, so I first need to convert the `TypedArray` to an image, then draw it again to resized it. I prefer to read it manually and put it into a `THREE.DataTexture`.

gchoqueux reviewed (commented):
It should be interesting to create a itowns plugin to support Tiff.

zarov committed:
feat(system): expose Capabilities class in Main.

Added a method to check the maximum texture size in the same time.

zarov commented:
Updated and rebased, I moved the parser to be a plugin. I also exposed the `Capabilities` class in the same time. In a next PR I'll start cleaning the `examples/` folder, because the plugins are hard to see.
